table/block_based/block.h

```diff
@@ -419,15 +421,16 @@ class BlockIter : public InternalIteratorBase<TValue> {
   int CompareCurrentKey(const Slice& other) {
     if (raw_key_.IsUserKey()) {
       assert(global_seqno_ == kDisableGlobalSequenceNumber);
-      return ucmp_wrapper_.Compare(raw_key_.GetUserKey(), other);
+      return ucmp().Compare(raw_key_.GetUserKey(), other);
     } else if (global_seqno_ == kDisableGlobalSequenceNumber) {
-      return icmp_.Compare(raw_key_.GetInternalKey(), other);
+      return icmp().Compare(raw_key_.GetInternalKey(), other);
     }
-    return icmp_.Compare(raw_key_.GetInternalKey(), global_seqno_, other,
-                         kDisableGlobalSequenceNumber);
+    return icmp().Compare(raw_key_.GetInternalKey(), global_seqno_, other,
+                          kDisableGlobalSequenceNumber);
```

Comment:
Hi @ajkr , I found that after this PR, each time when we call a CompareCurrentKey operation a new comparator will be initiated and constructed. I did some basic profiling and found that this may be a bottleneck. I was wondering what is the reason for this change, and would it harm performance? Thanks!

Reply:
Thanks for your reply. Could you explain more? From what I see, DataBlockIterator still calls CompareCurrentKey in BlockIter and either icmp() or ucmp() will allocate a new comparator.

Reply:
@mrambacher Thanks. I get that the "raw comparator" will be saved as raw_ucmp_ and passed as pointer to become UserComparatorWrapper when calling either icmp() or ucmp(), so this step has no allocation cost. However, calling icmp() or ucmp() will create a new Comparator instance, which is a subclass of Customizable. My concern is that the construction and destruction of such instances in every comparison operation may have a cost, as showed in some of my profiling.

Reply:
@XinyuZeng I was suggesting something different.  Store the "Internal Comparator" in the iterator (the icmp) instead of the raw comparator.  The icmp now returns this without creating anything.  The ucmp returns the "user_comparator" from the InternalKeyComparator -- again without creating anything.  The size of the object does not change (just storing a different pointer) and there is no overhead of creating objects on every call.

If you are not comfortable attempting this change, I can try to look into it more later in the week.

Reply:
@mrambacher Yes that is exactly what I am thinking about----I thought you were saying such logic already exists in code yesterday. I can attemp it this weekend and come back to you if I encounter any challenge since I did not commit to rocksdb before. Many thanks for your explaination!